source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -274,22 +274,48 @@ void Filter::UpstreamCallbacks::onEvent(Network::ConnectionEvent event) {
 }
 
 void Filter::UpstreamCallbacks::onAboveWriteBufferHighWatermark() {
-  ASSERT(!on_high_watermark_called_);
-  on_high_watermark_called_ = true;
-
-  if (parent_ != nullptr) {
-    // There's too much data buffered in the upstream write buffer, so stop reading.
-    parent_->readDisableDownstream(true);
+  if (above_high_watermark_counter_++ == 0) {
+    if (parent_ != nullptr) {
+      ENVOY_CONN_LOG(
+          trace,
+          "tcp_proxy disables downstream read because upstream write buffer is now above high "
+          "watermark. above_high_watermark_counter_ = {}",
+          *parent_->downstreamConnection(), above_high_watermark_counter_);
+      // There's too much data buffered in the upstream write buffer, so stop reading.
+      parent_->readDisableDownstream(true);
```

Comment:
the stats in read disable downstream assume calling disable pauses or resumes reading.
I'd suggest either only calling readDisabledownstream when this transitions, or updating readDisableDownstream to check internally.